### ISSUE TITLE
How do you transfer tokens other than SOL?

### ISSUE THREAD

**Issue by cyriimm** [michaelhly/solana-py#30]
How do you transfer tokens other than SOL? I have figured out how to transfer SOL , is there a way to send SRM?

**Comment by michaelhly** [comment]
Working on it in this PR: https://github.com/michaelhly/solana-py/pull/25.

Will get it merged soon.

**Comment by michaelhly** [comment]
Hi @cyriimm. Please update to v0.2.0 and you should be able to transfer SPL tokens now. 

Here is a snippet you can try:
```python
from solana.publickey import PublicKey
from solana.account import Account
from solana.rpc.api import Client
from solana.transaction import Transaction
from spl.token.instructions import transfer, TransferParams

serum = PublicKey(SRMuApVNdxXokk5GT7XD5cUUgXMBCoAz2LHeuAoKWRt) # Mainnet address
sender, dest = Account(), Account().  # These are arbitary accounts
transfer_params = TransferParams(
     amount=10000,
     dest=dest.public_key(),
     owner=sender.public_key(),
     program_id=serum,
     source=sender.public_key())
txn = transaction.add(transfer(transfer_params))
solana_client = Client("https://api.mainnet-beta.solana.com")
solana_client.send_transaction(txn, sender)
```

Let me know if you run into any issues, SPL-token related docs are over [here](https://michaelhly.github.io/solana-py/spl.html#spl-token).

**Comment by cyriimm** [comment]
`from solana.publickey import PublicKey
from solana.account import Account
from solana.rpc.api import Client
from solana.transaction import Transaction
from spl.token.instructions import transfer, TransferParams

serum = PublicKey("SRMuApVNdxXokk5GT7XD5cUUgXMBCoAz2LHeuAoKWRt") # Mainnet address
sender, dest = Account(), Account()

print(sender.public_key())# These are arbitary accounts
transfer_params = TransferParams(
     amount=10000,
     dest=dest.public_key(),
     owner=sender.public_key(),
     program_id=serum,
     source=sender.public_key())
txn = Transaction()
txn.add(transfer(transfer_params))
solana_client = Client("https://api.mainnet-beta.solana.com")
solana_client.send_transaction(txn, sender)
`

**Comment by michaelhly** [comment]
@cyriimm Can you post the code snippet you're using to send Serum tokens?

**Comment by cyriimm** [comment]
Below is the code snippet :)
`
from solana.publickey import PublicKey
from solana.account import Account
from solana.rpc.api import Client
from solana.transaction import Transaction
from spl.token.instructions import transfer, TransferParams

serum = PublicKey("SRMuApVNdxXokk5GT7XD5cUUgXMBCoAz2LHeuAoKWRt") # Mainnet address
sender, dest = Account(), Account()

print(sender.public_key())
#The sender key is not a SRM address but a SOL address

print(sender.public_key())# These are arbitary accounts
transfer_params = TransferParams(
     amount=2,
     dest=dest.public_key(),
     owner=sender.public_key(),
     program_id=serum,
     source=sender.public_key())
txn = Transaction()
txn.add(transfer(transfer_params))
solana_client = Client("https://api.mainnet-beta.solana.com")
solana_client.send_transaction(txn, sender)
`

**Comment by michaelhly** [comment]
> Below is the code snippet :)
> `
> from solana.publickey import PublicKey
> from solana.account import Account
> from solana.rpc.api import Client
> from solana.transaction import Transaction
> from spl.token.instructions import transfer, TransferParams
> 
> serum = PublicKey("SRMuApVNdxXokk5GT7XD5cUUgXMBCoAz2LHeuAoKWRt") # Mainnet address
> sender, dest = Account(), Account()
> 
> print(sender.public_key())
> #The sender key is not a SRM address but a SOL address
> 
> print(sender.public_key())# These are arbitary accounts
> transfer_params = TransferParams(
> amount=2,
> dest=dest.public_key(),
> owner=sender.public_key(),
> program_id=serum,
> source=sender.public_key())
> txn = Transaction()
> txn.add(transfer(transfer_params))
> solana_client = Client("https://api.mainnet-beta.solana.com")
> solana_client.send_transaction(txn, sender)
> `

FYI calling `Account()` will generate an account based on a random seed. I would recommend you store your private key as an environment variable and create an account as follow:

```python
sender = Account(os.environ['PRIVATE_KEY'])
```

**Comment by cyriimm** [comment]
for SRM are the 'mint' and 'token' addresses the same?
namely: SRMuApVNdxXokk5GT7XD5cUUgXMBCoAz2LHeuAoKWRt

**Comment by cyriimm** [comment]
Think I am getting somewhere with this :)
 I have now used spl token program address (that is spl)
When I try to send this transaction it says not signed correctly --> makes sense 
Does the initliazeAccount instruction need to be signed before sending in tx?
'
account, mint, owner, token = Account(4).public_key(),serum, Account(1).public_key(),spl
params = InitializeAccountParams(
    account=account,
    mint=mint,
    owner=owner,
    program_id=token,
)

sender=Account(1).public_key()
txn = Transaction()
txn.add(initialize_account(params))

solana_client = Client("https://api.mainnet-beta.solana.com")
solana_client.send_transaction(txn, sender)
'

**Comment by michaelhly** [comment]
Okay. We have all the functions in this library, but it's not very friendly yet. To create an spl-token account, we need to replicate these steps:

1. Get lamports needed to allocate memory for account with `get_minimum_balance_for_rent_exemption`.
2. Create an account via the SystemProgram: https://github.com/michaelhly/solana-py/blob/682f0f67015842b60b0edec5ca4e6b25d731e93c/solana/system_program.py#L290
3. Add the `initialize_account` instruction as we have tried.
4. Send transaction.

I'll be working on a token program module to abstract a lot of these steps away and also add tests. But if you'd like to experiment now, you can follow how the [js library](https://github.com/solana-labs/solana-program-library/blob/f5e0d793a9ff476e0ae767413955ffa3f8a6770c/token/js/client/token.js#L245) combines instructions together for common SPL-token operations.

**Comment by michaelhly** [comment]
A token client is now available in v0.05. Currently supports: `transfer`, `mint_to`, `get_balance`, `get_accounts` and `create_accounts`. I haven't gotten around to write proper docs, but hopefully, these tests would be good enough references for now:
https://github.com/michaelhly/solana-py/blob/ce34ff5195aa5ca4c051ce6359db0e08f04ced37/tests/integration/test_token_client.py#L16
